cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/TwillAppNaming.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.distributed;
+
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.common.base.Preconditions;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class to convert to and from the Twill application name for a {@link ProgramId}.
+ */
+public class TwillAppNaming {
+
+  // Pattern to split a Twill App name into [type].[namespaceId].[appName].[programName]
+  private static final Pattern APP_NAME_PATTERN = Pattern.compile("^(\\S+)\\.(\\S+)\\.(\\S+)\\.(\\S+)$");
+
+  private TwillAppNaming() { }
+
+  /**
+   * Returns the Twill app name to be used for a Twill app launched for a given program.
+   */
+  static String toTwillAppName(ProgramId programId) {
+    Id.Program program = programId.toId();
+    return String.format("%s.%s.%s.%s", program.getType().name().toLowerCase(),
+                         program.getNamespaceId(), program.getApplicationId(), program.getId());
+  }
+
+  /**
+   * Given a Twill app name, returns the ProgramId that was used to construct this Twill app name.
+   *
+   * @throws IllegalArgumentException if the given app name does not match the {@link TwillAppNaming#APP_NAME_PATTERN}.
+   */
+  public static ProgramId fromTwillAppName(String twillAppName) {
+    return fromTwillAppName(twillAppName, true);
+  }
+
+  /**
+   * Given a Twill app name, returns the ProgramId that was used to construct this Twill app name.
+   * Returns null if mustMatch is false, and if the specified Twill app name does
+   * not match the {@link TwillAppNaming#APP_NAME_PATTERN}.
```

Comment:
Right. Originally, I think the pattern was in a different class, so was fully qualifying. Will change.